cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -1828,6 +1828,120 @@ struct GroupedAllImpl : public GroupedAggregator {
   TypedBufferBuilder<bool> seen_;
 };
 
+// ----------------------------------------------------------------------
+// CountDistinct/Distinct implementation
+
+struct GroupedCountDistinctImpl : public GroupedAggregator {
+  Status Init(ExecContext* ctx, const FunctionOptions* options) override {
+    ctx_ = ctx;
+    pool_ = ctx->memory_pool();
+    return Status::OK();
+  }
+
+  Status Resize(int64_t new_num_groups) override {
+    num_groups_ = new_num_groups;
```

Comment:
@michalursa what do you think of adding `Grouper::Reserve(additional_capacity_hint)`?

Reply:
The keys here are pairs of `[value, group_id]` so just having the number of groups doesn't give much of a capacity hint.